File: src/main/java/com/zoarial/iot/dao/DAOHelper.java
package com.zoarial.iot.dao;

import javax.persistence.EntityManagerFactory;
import javax.persistence.Persistence;
import java.util.concurrent.atomic.AtomicReference;

public class DAOHelper {
    private final static AtomicReference<EntityManagerFactory> entityManagerFactory = new AtomicReference<>();

    private DAOHelper() {

    }

    public static EntityManagerFactory getEntityManagerFactory() {
        EntityManagerFactory emf = entityManagerFactory.getOpaque();
        if(emf == null) {
            synchronized (entityManagerFactory) {
                if(entityManagerFactory.getPlain() == null) {
                    emf = Persistence.createEntityManagerFactory("ZIoT");
                    entityManagerFactory.setPlain(emf);
                }
            }
        }
        return emf;
    }
    public static void setEntityManagerFactory(String name) {
        synchronized (entityManagerFactory) {
            entityManagerFactory.setPlain(Persistence.createEntityManagerFactory(name));
        }
    }
}

File: src/main/java/com/zoarial/iot/dao/DAOHelper.kt
package com.zoarial.iot.dao

import jakarta.persistence.EntityManagerFactory
import jakarta.persistence.Persistence
import java.util.concurrent.atomic.AtomicReference

object DAOHelper {
    private val entityManagerFactory = AtomicReference<EntityManagerFactory>()
    fun getEntityManagerFactory(): EntityManagerFactory {
        var emf = entityManagerFactory.opaque
        if (emf == null) {
            synchronized(entityManagerFactory) {
                if (entityManagerFactory.plain == null) {
                    emf = Persistence.createEntityManagerFactory("ZIoT")
                    entityManagerFactory.plain = emf
                }
            }
        }
        return emf
    }

    fun setEntityManagerFactory(name: String?) {
        synchronized(entityManagerFactory) { entityManagerFactory.setPlain(Persistence.createEntityManagerFactory(name)) }
    }
}